File: java/206_Reverse_Linked_List.java
# Iteration
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode new_head = null;
        while(head != null){
            ListNode temp = head.next;
            head.next = new_head;
            new_head = head;
            head = temp;
        }
        return new_head;
    }
}

# Recursion
class Solution {
    public ListNode reverseList(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }
        ListNode p = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return p;
    }
}

File: python/206_Reverse_Linked_List.py
# Recursion
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head
        new_head = reverseList(head.next)
        head.next.next = head
        head.next = None
        return new_head
       
# Iteration
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        new_head = None
        while head:
            temp = head.next
            head.next = new_head
            head, new_head = temp, head
        
        return new_head
